### PR TITLE
Update docs with new tensor repr

### PR DESCRIPTION
**Pull request by li-roy**
Update the docs as a follow up to #6370.

cc: @gchanan 

TODOs:
- [x] include things from the migration guide (dtype, device charts, etc). They're very nice
- [x] better docs for torch.layout
- [x] torch.dtype (aliases) & torch.layout page
- [x] link torch.no_grad stuff in torch.* docs page
- [x] torch.set_default_dtype docs
- [x] Tensor() -> torch.tensor() or torch.empty()
- [x] nll_loss and binary_cross_entropy examples do not run
- [x] move torch.dtype/device/cuda up (at least above torch.legacy)
- [x] LongTensor -> torch.tensor OR torch.empty, depending on use case
- [x] link arg types to torch.dtype/device (create actual (maybe dummy) class doc)

Not critical/going to pass on these until later date. These are a little hard to actually get to show up in sphinx.
- [ ] somehow include the big infographic on tensor constructors. Maybe make a new docs page?
- [ ] torch.default_generator docs missing

### CODE REVIEW

File: torch/__init__.py [hunk 1]
```diff
@@ -138,18 +138,15 @@ def set_default_tensor_type(t):
...
         >>> torch.Tensor([1.2, 3])
+        tensor([ 1.2000, 3.0000], dtype=torch.float64)
 
-         1.2000
-         3.0000
-        [torch.DoubleTensor of size (2,)]
-
+        >>> torch.set_default_tensor_type(torch.cuda.double)
+        >>> torch.Tensor([2, 3])
+        tensor([ 2, 3], dtype=torch.cuda.float64)
```
Comment (fmassa) [inline torch/__init__.py]: Aren't we separating device from dtype? Might be better to modify the print accordingly?

Comment (SsnL) [reply torch/__init__.py]: The device will be separated from dtype after #6470 . This PR needs to be updated when that merges...

Comment (li-roy) [reply torch/__init__.py]: Yeah, should be a small update because CPU will be the implicit, so only this example will need to be updated.

**Review by ezyang — DISMISSED** [review]
There is a spurious submodule update on gloo

File: docs/source/tensors.rst [hunk 2]
```diff
@@ -42,6 +42,14 @@ A tensor can be constructed from a Python :class:`list` or sequence using the
      1 -1
     [torch.FloatTensor of size (2,2)]
 
+.. warning::
+
+    :func:`torch.tensor` always copies :attr:`data`. If you have a Tensor
+    ``data`` and want to avoid a copy, use :func:`torch.Tensor.requires_grad_`
```
Comment (SsnL) [inline docs/source/tensors.rst]: @zou3519 Can you clarify here that users should use `.to` if want to change types/devices, and use these two methods if they want to change `requires_grad`.

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -6342,24 +5015,11 @@ def parse_kwargs(desc):
...
-    [torch.FloatTensor of size ()]
-    ,
-    -1.0402
-    [torch.FloatTensor of size ()]
-    )
-
+    (tensor(1), tensor(-0.4730))
 """)
 
 add_docstr(torch.fft,
```
Comment (SsnL) [inline torch/_torch_docs.py]: Would really appreciate if the `fft` docs below can be updated too. :)

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -6342,24 +5015,11 @@ def parse_kwargs(desc):
...
-    -0.3534
-    [torch.FloatTensor of size ()]
-
+    tensor(0.6231)
     >>> torch.logdet(A)
-
-    nan
-    [torch.FloatTensor of size ()]
-
+    tensor(-0.4730)
```
Comment (SsnL) [inline torch/_torch_docs.py]: It'd be nice if we can generate one that has negative `det`... The original purpose was to show that `slogdet` can handle negative `det`.

Comment (zou3519) [reply torch/_torch_docs.py]: We're making doc changes to li-roy/pytorch:tensorprintdocs -- feel free to push changes!

File: docs/source/tensors.rst (hunk 2 above)
Comment (zou3519) [reply docs/source/tensors.rst]: Yup!

Comment (gchanan) [reply docs/source/tensors.rst]: I feel like there is a phrase missing here.  What is the user trying to do in each case?

**Pull request merged**